statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/functions/FlinkStateDelayedMessagesBuffer.java

```diff
@@ -26,41 +33,121 @@
-          InternalListState<String, Long, Message> bufferState) {
+      @Label("delayed-messages-buffer-state") InternalListState<String, Long, Message> bufferState,
+      @Label("delayed-message-index") MapState<String, Long> messageIdToTimestamp) {
     this.bufferState = Objects.requireNonNull(bufferState);
+    this.messageIdToTimestamp = Objects.requireNonNull(messageIdToTimestamp);
   }
 
   @Override
-  public void add(Message message, long untilTimestamp) {
-    bufferState.setCurrentNamespace(untilTimestamp);
+  public void forEachMessageAt(long timestamp, Consumer<Message> fn) {
     try {
-      bufferState.add(message);
+      forEachMessageThrows(timestamp, fn);
     } catch (Exception e) {
-      throw new RuntimeException("Error adding delayed message to state buffer: " + message, e);
+      throw new IllegalStateException(e);
     }
   }
 
   @Override
-  public Iterable<Message> getForTimestamp(long timestamp) {
-    bufferState.setCurrentNamespace(timestamp);
-
+  public OptionalLong removeMessageByCancellationToken(String token) {
     try {
-      return bufferState.get();
+      return remove(token);
     } catch (Exception e) {
-      throw new RuntimeException(
-          "Error accessing delayed message in state buffer for timestamp: " + timestamp, e);
+      throw new IllegalStateException("Failed clearing a message with id " + token, e);
     }
   }
 
   @Override
-  public void clearForTimestamp(long timestamp) {
+  public void add(Message message, long untilTimestamp) {
+    try {
+      addThrows(message, untilTimestamp);
+    } catch (Exception e) {
+      throw new RuntimeException("Error adding delayed message to state buffer: " + message, e);
+    }
+  }
+
+  // -----------------------------------------------------------------------------------------------------
+  // Internal
+  // -----------------------------------------------------------------------------------------------------
+
+  private void forEachMessageThrows(long timestamp, Consumer<Message> fn) throws Exception {
     bufferState.setCurrentNamespace(timestamp);
+    for (Message message : bufferState.get()) {
+      removeMessageIdMapping(message);
+      fn.accept(message);
+    }
+    bufferState.clear();
+  }
+
+  private void addThrows(Message message, long untilTimestamp) throws Exception {
+    bufferState.setCurrentNamespace(untilTimestamp);
+    bufferState.add(message);
+    Optional<String> maybeId = message.cancellationToken();
+    if (!maybeId.isPresent()) {
+      return;
+    }
+    String messageId = maybeId.get();
+    @Nullable Long previousTimestamp = messageIdToTimestamp.get(messageId);
+    if (previousTimestamp != null) {
+      throw new IllegalStateException(
+          "Trying to associate a message with id "
+              + messageId
+              + " and timestamp "
+              + untilTimestamp
+              + ", but a message with the same id exists and with a timestamp "
```

Comment:
nit: excessive "and"

Reply:
Same here: `id` --> `cancellation token` in the error messages.
We seem to refer to these as cancellation tokens on the API surface.